### PR TITLE
WIP, ENH: add openblas runtime version inspection

### PR DESCRIPTION
**Pull request by tylerjereddy**
**Background**: The process of building wheels for different platforms can be a little hard to follow sometimes, and even inspecting a NumPy wheel to determine which version of `openblas` it has been linked against can be highly non-trivial. This very recently lead to the upstream merge [of a PR that adds a version number](https://github.com/xianyi/OpenBLAS/pull/1890) to an extension function in `openblas`. 

While we may eventually want to try to convince the `openblas` team to go a step farther and add the git commit hash for further clarity, we should be able to leverage the version string as a first pass verification & to provide the base infrastructure for runtime openblas version information inspection.

**Questions / Thoughts**:
1. we may want to provide direct access to the version information in `numpy.distutils.system_info.get_info('openblas')` which currently generates
a dictionary that might be expanded with a few new keys like this?
```python
{'libraries': ['openblas', 'openblas'], 
'library_dirs': ['/usr/lib'],
'language': 'c', 
'define_macros': [('HAVE_CBLAS', None)]
'version': '0.3.5dev', # should we add this?
'hash': something, # maybe some day?
}
```

2. We'd need a graceful way to handle the complete absence of openblas, and versions of openblas prior to the `0.3.4` release where the build info string does not include the version number. Basically, something flexible like the dictionary above is probably ok I guess.

3. In all likelihood the WIP code here is in the wrong files / paths -- should it all be diverted to distutils?

4. Do people want to be able to access the openblas version number in C-level code? I.e., make code path decisions in C based on openblas version? I suspect this would get ugly pretty fast given that we are not exclusive to supporting an openblas backend, but better to ask this now than to need it later & have diverging methods to inspect the openblas version info.

5. My current primary envisioned use case is quite simple -- in our development CI we have a variable i.e., OPENBLAS_VERSION -- this will pull in the appropriate wheel build URL, and then after we install NumPy from the assembled wheel we do a runtime sanity check that we have the correct linking (no interference from CI system changes, and provide absolute check / feedback to avoid sifting through complexities on MacPython wheels, etc. -- eventually perhaps upstreaming the check to post-mortem on wheels before they upload).

### CODE REVIEW

**Comment by rgommers** [comment]
> we may want to provide direct access to the version information in numpy.distutils.system_info.get_info('openblas') which currently generates
a dictionary that might be expanded with a few new keys like this?

Yes, I think that would be good. The new function added in this PR should be private I think.

(2) yes the dict should work fine I think

> In all likelihood the WIP code here is in the wrong files / paths -- should it all be diverted to distutils?

No, not in `numpy.distutils` imho. I think the openblas-specific C code fits best in `numpy.linalg`.

> Do people want to be able to access the openblas version number in C-level code?

No, unnecessary complexity. No one should be writing code even in Python, I think we want version numbers for diagnostic purposes.

File: numpy/linalg/setup.py [hunk 1]
```diff
@@ -53,8 +53,41 @@ def get_lapack_lite_sources(ext, build_dir):
...
+                             sources=['openblas_config.c'],
+                             extra_info=lapack_info,
+                             libraries=['openblas'],
+                             )
     return config
 
 if __name__ == '__main__':
     from numpy.distutils.core import setup
     setup(configuration=configuration)
+    import numpy.distutils.system_info
```
Comment (tylerjereddy) [inline numpy/linalg/setup.py]: this block never gets executed -- I need to move this code somewhere more appropriate; perhaps at some point of linalg initialization instead of in the distutils spaghetti

File: numpy/linalg/setup.py [hunk 2]
```diff
@@ -53,8 +53,41 @@ def get_lapack_lite_sources(ext, build_dir):
...
+        # as the docs clearly indicate
+        info = system_info.system_info.saved_results['openblas_info']
+        new_info = {}
+        if openblas_config_list[0] == b"OpenBLAS":
+            # version string will be present
+            new_info['version'] = openblas_config_list[1]
+        else:
+            # older OpenBLAS config API
+            new_info['version'] = None
+        info.update(new_info)
```
Comment (tylerjereddy) [inline numpy/linalg/setup.py]: the update won't happen in time to show the openblas version in the `FOUND` block one sees when distutils runs during setup, but it would show up if called later

I suppose this makes sense maybe -- if we're associating this machinery with linalg, then the libraries have to be detected first, and then the extension code is called to expand that library info with the version number.

Comment (tylerjereddy) [reply numpy/linalg/setup.py]: In other words, we need openblas to be initially detected by distutils so that this linalg extension can be built, and then build up the openblas lib info available to the former.

File: numpy/linalg/tests/test_openblasconfig.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+@pytest.mark.skipif(len(system_info.get_info('openblas')) < 2,
+                     reason="Requires openblas")
+def test_openblas_version_string():
+    # DEBUG only -- probing the config string
+    # in CI, where openblas >= 0.3.4 is available
+    # on Windows
+    from numpy.linalg import openblas_config
+    actual = openblas_config._openblas_info()
+    # currently the string looks like this on Windows Azure:
+    # b'OpenBLAS 0.3.5.dev DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=24'
```
Comment (tylerjereddy) [inline numpy/linalg/tests/test_openblasconfig.py]: lots of debug stuff here, but can confirm that we have access to the version string in my fork CI run, but I think the aim is to route this through distutils get_info dict retrieval as I've started doing

File: numpy/linalg/setup.py [hunk 4]
```diff
@@ -53,8 +53,41 @@ def get_lapack_lite_sources(ext, build_dir):
...
+        # numpy/distutils/system_info has already initialized
+        # at this stage, so just update the dictionary for
+        # openblas info
+        from numpy.linalg import openblas_config
+        openblas_config_str = openblas_config._openblas_info()
+        openblas_config_list = openblas_config_str.split()
+        # slightly obscure API for system_info objects;
+        # they weren't really intended for public modification
+        # as the docs clearly indicate
+        info = system_info.system_info.saved_results['openblas_info']
```
Comment (tylerjereddy) [inline numpy/linalg/setup.py]: messy, but only get_info is really supposed to be public, setting was never intended based on the comments in system_info

File: numpy/linalg/setup.py (hunk 1 above)
Comment (rgommers) [reply numpy/linalg/setup.py]: I suggest to remove that code block then, doesn't really seem useful.

Comment (rgommers) [reply numpy/linalg/setup.py]: And I'd prefer to have the mess contained in `numpy.distutils`. Also because then it's available to `scipy.linalg` and other users.

File: numpy/linalg/tests/test_openblasconfig.py (hunk 3 above)
Comment (rgommers) [reply numpy/linalg/tests/test_openblasconfig.py]: nice. and yes, I agree, in `get_info` is the right place

**Comment by tylerjereddy** [comment]
Looking a bit more solid now maybe -- Azure CI should pass with `OpenBLAS` version audits performed on Linux & Windows; Mac Azure doesn't use `OpenBLAS` yet so that'll have to wait.

I think Ralf wants me to move the code that adjusts the distutils dictionary containing the version data to somewhere within distutils itself. I'm not exactly certain where I'll put it, or how that will improve usage by other libraries (don't they need NumPy either way with the C code living in linalg?).

May be a nice touch if the dictionary was updated fast enough to display when distutils is actually building NumPy itself--discussed above--so the `FOUND` blocks show `version` right away, instead of a bit later in control flow.

**Comment by rgommers** [comment]
> I think Ralf wants me to move the code that adjusts the distutils dictionary containing the version data to somewhere within distutils itself. I'm not exactly certain where I'll put it, or how that will improve usage by other libraries (don't they need NumPy either way with the C code living in linalg?).

It seems to me that that C code should also go into `numpy.distutils`. It's similar to some of the conditional compiles we do there for checking availability of headers etc. If it goes into `numpy.linalg`, then every other package that uses OpenBLAS needs to reinvent this particular wheel.

**Comment by tylerjereddy** [comment]
I've spent a fair bit of time on this now and I still don't see a clear route to silo this functionality into `distutils`. Perhaps others have advice. There are a few reasons for this:

- NumPy itself builds `core` (or at least uses parts of it) before `distutils` system_info is fully set up and some of the dummy C code build / header stuff is in `numpy/core/src`, so another package would need that to compile anyway; `core` building uses `system_info` but I want `system_info` to have access to compiled code in `distutils`
- `distutils` also depends on `numpy.lib` for some things like `get_include()` -- `distutils` just doesn't seem to be designed as a true silo at the moment; yes, its true that projects like SciPy do git pull the latest version of distutils in isolation for some CI build scenarios, but it is still placed in the context (replaces old distutils path) of the rest of NumPy, and so fundamentally still depends on other machinery.

Whenever I try to completely isolate the functionality to distutils I end up with a "chicken & egg problem" because if NumPy distutils is truly an independent utility for package building, then NumPy certainly isn't treating it as such.

`distutils` doesn't even have a `src` directory for C code at the moment -- are we ready to add that in and have distutils compile itself?

Is it really a good idea for me to try to refactor / silo distutils?

**Comment by rgommers** [comment]
> Is it really a good idea for me to try to refactor / silo distutils?

Okay, if it's that hard then probably not a good idea / not worth the effort. Whatever works then I guess, we can copy it into SciPy then if it's useful ...

**Comment by tylerjereddy** [comment]
We could put the C source in a `distutils` path, and maybe even the Python interface that accesses it (in a function that only works when the C extension is built).

But both NumPy and downstream libraries would still be responsible for writing a `setup.py` block oustide `distutils` that builds the C extension with something like this:

```python
config.add_extension('openblas_config',
                     sources=['openblas_config.c'],
                     extra_info=lapack_info,
                     libraries=['openblas'],
                     )
```

There are certainly cases where a new version of `distutils` is copied into an older version of NumPy, but in those cases my understanding is that it is primarily pure Python code paths in `distutils` that are being leveraged.

Still, this approach would at least provide the raw low-level code needed to generate the interface, and maybe that's about as good as we can do for re-usability with current constraints.

**Comment by rgommers** [comment]
> There are certainly cases where a new version of distutils is copied into an older version of NumPy,

That's only an unfortunate workaround for when `numpy.distutils` contains new fixes that are needed for, e.g., SciPy and SciPy cannot drop support for older NumPy versions that miss the fix. It wouldn't be necessary if `numpy.distutils` was a standalone package, but that boat has sailed a long time ago.

**Comment by tylerjereddy** [comment]
The somewhat-related discussion in https://github.com/scipy/scipy/issues/9659 now has me thinking about the possibility of using something like `try_link` (or similar) from vanilla distutils. If it isn't too fragile, might be a nice way to "inline compile" the small body of C code needed to yank out the version number.

I'll have to play with it a bit I guess.

File: numpy/linalg/__init__.py [hunk 5]
```diff
@@ -50,6 +50,28 @@
...
+    info = sinfo.system_info.saved_results['openblas_info']
+    new_info = {}
+    if openblas_config_list[0] == b"OpenBLAS":
+        # version string will be present
+        new_info['version'] = openblas_config_list[1]
+    else:
+        # older OpenBLAS config API
+        new_info['version'] = None
+    info.update(new_info)
+
```
Comment (mattip) [inline numpy/linalg/__init__.py]: Could you replace this with a ctypes-based check in [`openblas_info.calc_info`](https://github.com/numpy/numpy/blob/v1.16.0rc2/numpy/distutils/system_info.py#L1756)? You would need to find the exact shared object name sucessfully found in `self.check_libs`, but then the rest is easy (assuming ctypes exists, which I think is true for any python version we care about)
```
dll = ctypes.CDLL(lib_name)
openblas_get_config = dll.openblas_get_config
openblas_get_config.restype = ctypes.c_char_p
openblas_config_str = openblas_get_config()
```

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: I'm not sure I understand how this helps. You're certainly right that it would be nice to intervene in the `system_info.py` module, and even better so if the initial detection of `OpenBLAS` is immediately flagged / logged with the version during the NumPy build.

But the problem is compiling the C module into the NumPy namespace inline in the Python code in `system_info.py`, or somewhere upstream in the build process, and I think this Python code is used extremely early in the NumPy build control flow. It is harder to go much farther upstream in the build process than the detection of the libraries though, which I think is why it is just so much easier to compile from a NumPy submodule "downstream" of distutils in the build process.

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: In other words, in your suggestion, how do I generate the shared object file upstream in the workflow? Maybe just the `try_link` or `try_compile` stuff still is a possibility if I can hook into the return values.

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: Ah, cool, you're referring to openblas itself instead of a compiled file associated with the new C extension. I should definitely try that since compiling the new C extension is the major bottleneck.

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: Seems to work well in my hands for Mac / Linux, but for Windows I'm having to think a bit more because we link with the static library there. Maybe I can cheat and use the dynamic libraries that come in the bundle just for the versioning purpose.

Comment (mattip) [reply numpy/linalg/__init__.py]: Strange, it seems the windows wheels for 1.15.4 dynamically link to an OpenBLAS dll. Are the wheels different from the dev builds?

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: https://github.com/MacPython/numpy-wheels/blob/master/.appveyor.yml#L89

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: That's the static lib used for wheels too -- we even physically remove the dynamic libs for good measure

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: I think there's a good reason for that too -- I think it is related to Windows + mingw + openblas combination and static being required for some reason. Not sure if that is still true, but pretty sure we still do it, and changing that in the openblas wheel generation workflow could cause quite a bit of disruption.

There is a `dll.a` file available as well on Windows, but so far `ctypes` doesn't like it either, and I think those may just be helper files for when the true dynamic lib is available.

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: So, I think it is back to the original issue -- fully portable detection of static / dynamic openblas lib version contained to distutils is hard. Using C extension is more portable, but hard to confine to distutils.

Comment (mattip) [reply numpy/linalg/__init__.py]: Well, when I download the [wheel for 1.16rc2](https://files.pythonhosted.org/packages/8a/51/44662dbc1963cd1e542459aa75e212cb9f1783b84b88192d33bf0d00b78c/numpy-1.16.0rc2-cp27-cp27m-win_amd64.whl) and unzip it, I find `numpy/.libs/libopenblas.IPBC74C7KURV7CB2PKT5Z5FNR3SIBV4J.gfortran-win_amd64.dll`. 

Using `depends.exe` I see that `numpy/core/_multiarray_umath.pyd` depends on that `dll`.

Perhaps the result of `self.check_libs` gives the import library rather than the dll, since MSVC links to a `*.lib` file not a `*.dll` file. At runtime there is an algorithm to determine where to fine the matching `*.dll`. listed inside the import library

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: The real question for me is whether the Windows openblas dynamic library is available before NumPy even starts compiling. So from our Azure openblas zip URL I download from here and unzip: https://3f23b170c54c2533c070-1c8a9b3114517dc5fe17b7c3f8c63a43.ssl.cf2.rackcdn.com/openblas-v0.3.3-186-g701ea883-win_amd64-gcc_7_1_0.zip

I don't think anything in there was usable with `cytpes` for dynamic library interaction in my hands. If there's an intermediate command needed to generate the requisite ctypes-tractable file first then I'd be glad to know it.

@matthew-brett Do you know offhand if these Windows openblas zip files are supposed to have dynamic libraries ready for i.e., ctypes access?

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: I certainly saw way more files for the MacOS version of the ZIP--in fact the code I wrote to parse for version number goes through the various files available in the zip and ignores the ones that don't work with ctypes--only a few of the files for Mac worked, and none of the Windows ones.

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: I do see `libopenblas_v0.3.3-186-g701ea883-gcc_7_1_0.dll` actually -- it is hiding in the `bin` path instead of `lib`, so I can play with that maybe!

Comment (mattip) [reply numpy/linalg/__init__.py]: The [azure build script](https://github.com/numpy/numpy/blob/v1.16.0rc2/azure-pipelines.yml#L160) only copies the `*.a` file from the zip. I wonder where the `/libopenblas.IPBC74C7KURV7CB2PKT5Z5FNR3SIBV4J.gfortran-win_amd64.dll` file that is in the wheel comes from?

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: @mattip Does that work for you for the DLL in the openblas zip file mentioned above?

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: on a Windows machine

Comment (mattip) [reply numpy/linalg/__init__.py]: @tylerjereddy: yes

Comment (tylerjereddy) [reply numpy/linalg/__init__.py]: even `depends.exe` used in combination with the absolute file path or in the same directory as the OpenBLAS DLL produces "The system cannot find the file specified (2)." after Dependency Walker loads.

I need a way to get reliable missing DLL / dependencies on the openblas DLL I download, irrespective of my current path / other system configs, assuming this is the problem.

**Comment by rgommers** [comment]
@tylerjereddy was it this PR or gh-12790 that you wanted review for? and did you try the `try_link` discussed above?

**Comment by tylerjereddy** [comment]
The other one you linked. I'll close this for now to avoid confusion.
